sdk/go/pulumi/types.go

```diff
@@ -1248,3 +1248,64 @@ func init() {
 	RegisterOutputType(ResourceOutput{})
 	RegisterOutputType(ResourceArrayOutput{})
 }
+
+// coerceTypeConversion assigns src to dst, performing deep type coercion as necessary.
+func coerceTypeConversion(src interface{}, dst reflect.Type) (interface{}, error) {
+	makeError := func(src, dst reflect.Value) error {
+		return fmt.Errorf("expected value of type %s, not %s", dst.Type(), src.Type())
+	}
+	var coerce func(reflect.Value, reflect.Value) error
+	coerce = func(src, dst reflect.Value) error {
```

Comment:
Good recursive algorithm 👍 

Contravariance implemented in ApplyT is extremely general.

Reply:
Just a query, can `any` not be used in place of `interface{}`? Looks cleaner, but I'm guessing it is for backward compatibility.

Reply:
We're just using go's built in `reflect.Type` `fmt.Stringer` implementation. I was going for an error message that was similar to what we had before (`i.(newType)`). I would assume that the go dev team prefers `interface {}` over `any` for backwards compatibility, but you'd need to ask them to be sure.

Reply:
@Graham-Beer we could use `any` here if we updated go.mod to require 1.18 and above. I believe we'll be doing that soon anyhow as 1.17 is no longer supported.